dockstore-cli-integration-testing/src/test/java/io/dockstore/client/cli/BasicIT.java

```diff
@@ -1498,6 +1498,39 @@ public void testRefreshingUserMetadata() {
         Assert.assertEquals("One user should have this info now, there are " + count, 1, count);
     }
 
+    @Test
+    public void launchToolChecksumValidation() {
+        // manual publish the tool
+        Client.main(new String[] { "--config", ResourceHelpers.resourceFilePath("config_file.txt"), "tool", "publish", "--entry",
+            "quay.io/dockstoretestuser/test_input_json", "--script" });
+
+        // refresh the tool
+        Client.main(new String[] { "--config", ResourceHelpers.resourceFilePath("config_file.txt"), "tool", "refresh", "--entry",
+            "quay.io/dockstoretestuser/test_input_json", "--script" });
+
+        // launch the tool
+        systemOutRule.clearLog();
+        Client.main(new String[] { "--config", ResourceHelpers.resourceFilePath("config_file.txt"), "tool", "launch", "--entry",
+            "quay.io/dockstoretestuser/test_input_json", "--json", ResourceHelpers.resourceFilePath("tool_hello_world.json"), "--script" });
+        assertTrue("Output should indicate that checksums have been validated",
+            systemOutRule.getLog().contains("Validated checksums") && !systemOutRule.getLog().contains("Cannot validate the checksum of the locally downloaded descriptor."));
+
+        // TODO: Currently, if a checksum is null the user is presented with a warning instead of throwing an exception. This should be fixed later to be more rigid and error out once checksums are more common.
+        testingPostgres.runUpdateStatement("UPDATE sourcefile SET checksums = '' WHERE path='/Dockstore.cwl';");
+
+        systemOutRule.clearLog();
+        Client.main(new String[] { "--config", ResourceHelpers.resourceFilePath("config_file.txt"), "tool", "launch", "--entry",
+            "quay.io/dockstoretestuser/test_input_json", "--json", ResourceHelpers.resourceFilePath("tool_hello_world.json"), "--script" });
+        assertTrue("Output should indicate that checksums have been validated",
+            systemOutRule.getLog().contains("Validated checksums") && systemOutRule.getLog().contains("Cannot validate the checksum of the locally downloaded descriptor."));
+
+        testingPostgres.runUpdateStatement("UPDATE sourcefile SET checksums = 'SHA-1:VeryFakeChecksum' WHERE path='/Dockstore.cwl';");
+
+        systemExit.expectSystemExitWithStatus(Client.API_ERROR);
```

Comment:
Wonder if you can test the exit message (Cannot validate the checksum ...) with https://stefanbirkner.github.io/system-rules/apidocs/org/junit/contrib/java/lang/system/ExpectedSystemExit.html#checkAssertionAfterwards-org.junit.contrib.java.lang.system.Assertion-

Reply:
Yes! Now, after the expected system exit, the expected output is [checked](https://github.com/dockstore/dockstore-cli/blob/7f6c85a4a186819af1abeaf2f795939c1fb693b4/dockstore-cli-integration-testing/src/test/java/io/dockstore/client/cli/GeneralWorkflowIT.java#L1143). Used for Tool test as well.